syncer/ddl_test.go

```diff
@@ -358,3 +370,156 @@ func (s *testSyncerSuite) TestResolveGeneratedColumnSQL(c *C) {
+		c.Assert(err, IsNil)
+		c.Assert(sqls, HasLen, 0)
+		c.Assert(tables, HasLen, 0)
+		sql = fmt.Sprintf("RENAME TABLE `test`.`%s` TO `test`.`t1`", ca.ghostname)
+		stmt, err = p.ParseOneStmt(sql, "", "")
+		c.Assert(err, IsNil)
+		sqls, tables, err = syncer.resolveDDLSQL(tctx, p, stmt, "test")
+		c.Assert(err, IsNil)
+		c.Assert(sqls, HasLen, 1)
+		c.Assert(sqls[0], Equals, newSQL)
+		tableName := &filter.Table{Schema: "test", Name: ca.ghostname}
+		c.Assert(tables, DeepEquals, map[string]*filter.Table{tableName.String(): tableName})
+	}
+}
+
+func (s *testSyncerSuite) TestDropSchemaInSharding(c *C) {
+	var (
+		targetDB  = "target_db"
+		targetTbl = "tbl"
+		sourceDB  = "db1"
+		source1   = "`db1`.`tbl1`"
+		source2   = "`db1`.`tbl2`"
+		tctx      = tcontext.Background()
+	)
+	clone, _ := s.cfg.Clone()
+	clone.ShardMode = config.ShardPessimistic
+	syncer := NewSyncer(clone, nil)
+	_, _, _, _, err := syncer.sgk.AddGroup(targetDB, targetTbl, []string{source1}, nil, true)
+	c.Assert(err, IsNil)
+	_, _, _, _, err = syncer.sgk.AddGroup(targetDB, targetTbl, []string{source2}, nil, true)
+	c.Assert(err, IsNil)
+	c.Assert(syncer.sgk.Groups(), HasLen, 2)
+	c.Assert(syncer.dropSchemaInSharding(tctx, sourceDB), IsNil)
+	c.Assert(syncer.sgk.Groups(), HasLen, 0)
+}
+
+type mockOnlinePlugin struct {
+	toFinish map[string]struct{}
+}
+
+func (m mockOnlinePlugin) Apply(tctx *tcontext.Context, tables []*filter.Table, statement string, stmt ast.StmtNode) ([]string, string, string, error) {
+	return nil, "", "", nil
+}
+
+func (m mockOnlinePlugin) Finish(tctx *tcontext.Context, schema, table string) error {
+	k := schema + table
+	if _, ok := m.toFinish[k]; !ok {
+		return errors.New("finish table not found")
+	}
+	delete(m.toFinish, schema+table)
+	return nil
+}
+
+func (m mockOnlinePlugin) TableType(table string) TableType {
+	return ""
+}
+
+func (m mockOnlinePlugin) RealName(schema, table string) (string, string) {
+	return "", ""
+}
+
+func (m mockOnlinePlugin) ResetConn(tctx *tcontext.Context) error {
+	return nil
+}
+
+func (m mockOnlinePlugin) Clear(tctx *tcontext.Context) error {
+	return nil
+}
+
+func (m mockOnlinePlugin) Close() {
+	return
+}
+
+func (s *testSyncerSuite) TestClearOnlineDDL(c *C) {
+	var (
+		targetDB  = "target_db"
+		targetTbl = "tbl"
+		source1   = "`db1`.`tbl1`"
+		key1      = "db1tbl1"
+		source2   = "`db1`.`tbl2`"
+		key2      = "db1tbl2"
+		tctx      = tcontext.Background()
+	)
+	clone, _ := s.cfg.Clone()
+	clone.ShardMode = config.ShardPessimistic
+	syncer := NewSyncer(clone, nil)
+	mock := mockOnlinePlugin{
+		map[string]struct{}{key1: {}, key2: {}},
+	}
+	syncer.onlineDDL = mock
+
+	_, _, _, _, err := syncer.sgk.AddGroup(targetDB, targetTbl, []string{source1}, nil, true)
+	c.Assert(err, IsNil)
+	_, _, _, _, err = syncer.sgk.AddGroup(targetDB, targetTbl, []string{source2}, nil, true)
+	c.Assert(err, IsNil)
+
+	c.Assert(syncer.clearOnlineDDL(tctx, targetDB, targetTbl), IsNil)
+	c.Assert(mock.toFinish, HasLen, 0)
```

Comment:
It will always `HasLen` `0` because we never store anything to it?

Reply:
it's in line 514, "finish" will remove one table from this member.